### ISSUE TITLE
Monorepo

### ISSUE THREAD

**Issue by berezovskyi** [eclipse/lyo#14]
We shall make a monorepo of all Lyo projects that are deployed to Maven (so, not including the test suites, RIOs, and Lyo Designer)

### Main code migration

- [x] core
- [x] client
- [x] domains
- [x] server
- [x] validation
- [x] trs-client
- [x] trs-server
- [x] store
- [ ] ldp
- [ ] rio

### Other migration aspects

- [x] Eclipse Jenkins build and internal deploy tested
- [x] POM settings for signing profile for Maven central
- [x] 🚨 Migrate issues
- [x]  🚨 Resolve critical issues in POMs, see #20 
- [x] Jenkins javadoc deployment (blocked by the POM issues with fulll javadoc gen)
- [ ] Merge/clean up READMEs
- [ ] Add notices to the old repos

### Misc / Low prio

- [x] Clean up old Jenkins configs
- [ ] Use a Jenkinsfile
- [ ] Set up Jenkins to build on push
- [x] validation build is slow
- [ ] https://github.com/OSLC/lyo-releng (we shall remove it afterwards)
- [ ] split various Lyo samples instead of batching them together - any small issue can fail the build
- [ ] update folder structure, eg move server artifacts under oauth/ folder

**Comment by berezovskyi** [comment]
I think we will move this up to 4.1.

The main reason is that it takes me around 10 PRs to release Lyo (even a milestone build), Jenkins CI is building cascaded jobs and I have to merge those 10 PRs in particular order. The tools for automating this improved release time from 4-6h to approx 1h but it's still not good enough.

Another reason is that changes break other repos and leave the whole Lyo in "unreleasable" state, e.g. https://github.com/eclipse/lyo.trs-client/runs/1380568111

But the change will be limited just to a monorepo, no Gradle move. And lyo-samples split will also not be included in this yet.

Proposed repos/folders that will be included:

```
- core (lyo.core)
- clients (lyo.client)
- domains (lyo.domains)
- lib
   \- store (lyo.store)
   \- trs
       \- trs-client
       \- trs-server
   \- oauth (lyo.server)
   \- validation (lyo.validation)
- misc 
   \- server
      \- registry
      \- wink
   \- rio (without samples)
   \- ldp (only server)
```

This will allow to make a release (eg RC release candidate I wanted to make for Lyo 4.0 after the new client migration) in just 2 PRs and prevent the case where trs-client breaks after lyo.client is updated.

@jadelkhoury LyoD is not included in this, that's why we are expecting 2 PRs to be needed for a release which I think fine with you.

**Comment by berezovskyi** [comment]
On the other hand, I think https://github.com/eclipse/lyo/issues/17 would require a bit more careful testing so making an RC release will be good for 4.0 and maybe we shall execute this migration now and not wait for 4.0. The only thing we shall take care of is all the PR shall be merged or closed before the move.

**Comment by berezovskyi** [comment]
Ah and yes, library version mismatch will be a thing of the past

**Comment by berezovskyi** [comment]
@jadelkhoury just make a sample monorepo with core and client merged into eclipse/lyo (this repo) here: https://github.com/berezovskyi/lyo-monorepo

Note that the tags are intact but will be bit tricky to work with: https://github.com/berezovskyi/lyo-monorepo/tags

**Comment by berezovskyi** [comment]
I used this to merge histories: https://stackoverflow.com/a/10548919/464590

I had to use https://gist.github.com/knu/111055 to rename non-core repo tags as their name clashed (in the real migration I will also rename the core tags from x.x.x to vx.x.x-core just as client)

**Comment by berezovskyi** [comment]
And of course POM files have not been fixed yet but I will do that once I hear no major objections

**Comment by berezovskyi** [comment]
Here is the script so far: https://gist.github.com/berezovskyi/a95ca0eb0d269c17a9f80229c24844f5

**Comment by berezovskyi** [comment]
I will proceed with the migration @jadelkhoury @jamsden hearing no major objections.

**Comment by jadelkhoury** [comment]
I don’t really know what this is about.
If this is related to the release process, then I have little to object on because I don’t know how it works anyway. 

If it is related to the code, then I think we have introduced many late changes just before releasing 4.0.0. Then maybe it should also wait.

**Comment by berezovskyi** [comment]
No code changes, just all folders in one repo. Thanks for replying.

--
Cheers,
Andrew

**Comment by berezovskyi** [comment]
@jadelkhoury @jamsden what do you think if I give up a bit on publishing multiple Javadocs and instead publish them all together:

![image](https://user-images.githubusercontent.com/64734/101266772-84376180-3752-11eb-82f1-2a26a7589c71.png)

Theoretically, we can do https://maven.apache.org/plugins/maven-javadoc-plugin/examples/group-configuration.html when we have extra time.

**Comment by jadelkhoury** [comment]
From an end-user, it is much better to have a single java-doc.
Even for a developer, now that we have all under 1 repo, 1 java-doc is easier.
This work (not just java-doc) will make our documentation so much easier.

**Comment by berezovskyi** [comment]
Thanks Jad, it is definitely better to have a single Javadoc for me, for example, you could see deprecated oslc-java-client-resources and new lyo-domains classes together. I was unsure if it would overwhelm new users. 

I will edit the Jenkins build then to publish this single Javadoc.

**Comment by berezovskyi** [comment]
@berezovskyi the javadoc is live at https://download.eclipse.org/lyo/docs/all/latest/apidocs/

**Comment by berezovskyi** [comment]
@jamsden @jadelkhoury looking at the task list at the top and the milestone https://github.com/eclipse/lyo/issues?q=is%3Aopen+is%3Aissue+milestone%3A4.0.0-RC and I think we are ready to release 4.0.0.RC. What do you say?

**Comment by jamsden** [comment]
Go!

**Comment by berezovskyi** [comment]
With https://github.com/eclipse/lyo/issues/80 fixed, I consider a monorepo migration complete.
